### PR TITLE
contravariant coyoneda

### PR DESCRIPTION
**Pull request by tpolecat**
This is mostly a copy/paste job from `Coyoneda` with the arrows flipped.

### CODE REVIEW

File: free/src/main/scala/cats/free/ContravariantCoyoneda.scala [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+   */
+  private[cats] def unsafeApply[F[_], A, B](fa: F[A])(ks0: List[Any => Any]): Aux[F, B, A] =
+    new ContravariantCoyoneda[F, B] {
+      type Pivot = A
+      val ks = ks0
+      val fi = fa
+    }
+
+  /** `ContravariantCoyoneda[F, ?]` provides a conntravariant functor for any `F`. */
+  implicit def catsFreeCovariantFunctorForCovariantCoyoneda[F[_]]: Contravariant[ContravariantCoyoneda[F, ?]] =
```
Comment (iravid) [inline free/src/main/scala/cats/free/ContravariantCoyoneda.scala]: Should be `Contravariant` here

File: free/src/main/scala/cats/free/ContravariantCoyoneda.scala [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  private[cats] val ks: List[Any => Any]
+
+  /** The composed transformer function, to be lifted into `F` by `run`. */
+  final def k: A => Pivot = Function.chain(ks)(_).asInstanceOf[Pivot]
+
+  /** Converts to `F[A]` given that `F` is a contravariant functor */
+  final def run(implicit F: Contravariant[F]): F[A] = F.contramap(fi)(k)
+
+  /** Converts to `G[A]` given that `G` is a contravariant functor */
+  final def foldMap[G[_]](trans: F ~> G)(implicit G: Contravariant[G]): G[A] =
```
Comment (kailuowang) [inline free/src/main/scala/cats/free/ContravariantCoyoneda.scala]: can we add a test for this one as well?

**Comment by kailuowang** [comment]
@tpolecat would you like to bring this one into the upcoming 1.1 release?

**Comment by tpolecat** [comment]
Yeah I'll try to do that test this morning.

**Comment by tpolecat** [comment]
I added a kind of lame test that exercises `foldMap` … I couldn't come up with a contravariant functor that was anything other than some wrapping of `? => A`. Happy to take suggestions but I think this may just be the nature of contravariant functors.

**Comment by kailuowang** [comment]
@iravid any other feedback on this one?

**Comment by iravid** [comment]
Nope 👍🏻

--
Itamar

**Comment by LukaJCB** [comment]
I'm very similar to Kai, the code looks very much okay to me, but I don't quite grasp the use cases, so if we could get one other review, maybe that'd be best.

**Comment by tpolecat** [comment]
As a practical matter the main use case is factoring out the boilerplate of writing a [stack-safe] contravariant functor. Simple example [here](https://github.com/tpolecat/doobie-tsql/blob/master/modules/core/src/main/scala/tsql/Write.scala#L30-L41).

**Pull request merged**